src/MessagePack/Internal/UnsafeRefSerializeHelper.cs

```diff
@@ -0,0 +1,1572 @@
+﻿// Copyright (c) All contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Buffers.Binary;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+#if NET8_0_OR_GREATER
+using System.Runtime.Intrinsics;
+#endif
+
+namespace MessagePack.Internal;
+
+internal static class UnsafeRefSerializeHelper
+{
+    internal static void Serialize(ref MessagePackWriter writer, ref bool input, int length)
```

Comment:
Is there any reason this signature couldn't be `ReadOnlySpan<bool>` instead of `ref bool, int`?

Reply:
[Vector128.Load](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.intrinsics.vector128.loadunsafe?view=net-8.0#system-runtime-intrinsics-vector128-loadunsafe-1(-0@)) method requires unsafe `T*` or `ref T`.
There is [MemoryMarshal.GetArrayDataReference](https://learn.microsoft.com/ja-jp/dotnet/api/system.runtime.interopservices.memorymarshal.getarraydatareference?view=net-8.0#system-runtime-interopservices-memorymarshal-getarraydatareference(system-array)) since .NET6.
It seemed  not so performant when go through `ReadOnlySpan&lt;T&gt;` in order to get `ref T` from `T[]`.

Reply:
Thanks for the thorough explanation. I suggest capturing it in code comments at the top of the file so that I or others don't go through later to "fix up the API", ignorant to the subtle but important perf impact your choices had.

Reply:
I wrote it.